include/flang/Lower/PFTBuilder.h

```diff
@@ -104,62 +75,48 @@ using ConstructStmts = std::tuple<
+               parser::CriticalConstruct, parser::DoConstruct,
+               parser::IfConstruct, parser::SelectRankConstruct,
+               parser::SelectTypeConstruct, parser::WhereConstruct,
+               parser::ForallConstruct, parser::CompilerDirective,
+               parser::OpenMPConstruct, parser::OmpEndLoopDirective>;
+
 template <typename A>
 static constexpr bool isActionStmt{common::HasMember<A, ActionStmts>};
 
 template <typename A>
-static constexpr bool isConstruct{common::HasMember<A, Constructs>};
+static constexpr bool isOtherStmt{common::HasMember<A, OtherStmts>};
 
 template <typename A>
 static constexpr bool isConstructStmt{common::HasMember<A, ConstructStmts>};
 
 template <typename A>
-static constexpr bool isOtherStmt{common::HasMember<A, OtherStmts>};
-
-template <typename A>
-static constexpr bool isGenerated{std::is_same_v<A, CGJump>};
+static constexpr bool isConstruct{common::HasMember<A, Constructs>};
 
 template <typename A>
 static constexpr bool isFunctionLike{common::HasMember<
     A, std::tuple<parser::MainProgram, parser::FunctionSubprogram,
                   parser::SubroutineSubprogram,
                   parser::SeparateModuleSubprogram>>};
 
-/// Function-like units can contains lists of evaluations.  These can be
+using LabelSet = llvm::SmallSet<parser::Label, 5>;
+using SymbolLabelMap = std::map<semantics::Symbol *, LabelSet>;
+
+/// Function-like units can contain lists of evaluations.  These can be
 /// (simple) statements or constructs, where a construct contains its own
 /// evaluations.
 struct Evaluation {
-  using EvalTuple = common::CombineTuples<ActionStmts, OtherStmts, Constructs,
-                                          ConstructStmts>;
+  using EvaluationTuple = common::CombineTuples<ActionStmts, OtherStmts,
+                                                ConstructStmts, Constructs>;
 
   /// Hide non-nullable pointers to the parse-tree node.
   template <typename A>
   using MakeRefType = const A *const;
-
-  using EvalVariant =
-      common::CombineVariants<common::MapTemplate<MakeRefType, EvalTuple>,
-                              std::variant<CGJump>>;
-
-  /// General ctor
-  template <typename A>
-  Evaluation(const A &a, const ParentType &p, const parser::CharBlock &pos,
-             const std::optional<parser::Label> &lab)
-      : u{&a}, parent{p}, pos{pos}, lab{lab} {}
-
-  /// Compiler-generated jump
-  Evaluation(const CGJump &jump, const ParentType &p)
-      : u{jump}, parent{p}, cfg{CFGAnnotation::Goto} {}
-
-  /// Construct ctor
-  template <typename A>
-  Evaluation(const A &a, const ParentType &parent) : u{&a}, parent{parent} {
-    static_assert(pft::isConstruct<A>, "must be a construct");
-  }
-
-  Evaluation() = delete;
-  Evaluation(const Evaluation &) = delete;
-  Evaluation(Evaluation &&) = default;
-
+  using EvaluationVariant = common::CombineVariants<
+      common::MapTemplate<MakeRefType, EvaluationTuple>>;
```

Comment:
It might be good to add a comment as to what this nested template majik is doing. (`MapTemplate` uses "map" in the sense of a functional language's "apply to each element" rather than "map" in the STL sense of a data structure. It's not clear this will be obvious to a passerby.)

Reply:
I think this is Jean's magic.  I think I'll need a concrete suggestion for a comment.

Reply:
Thanks wizard!  I've added this comment.  If I placed it in the wrong spot, or something I did could otherwise be improved, let me know and I'll update it.  Or make modifications yourself the next time you touch the code.
Thanks.